File: src/encoder.ts
/**
 * NEON - Neural Efficient Object Notation
 * Advanced Encoder Implementation v2.0
 *
 * Optimized for AI/LLM processing with maximum compression
 */

import {
  NeonValue,
  NeonObject,
  NeonArray,
  NeonEncodeOptions,
  NeonSchema,
  NeonSchemaField,
  NeonStats,
  NeonFieldType,
  DEFAULT_ENCODE_OPTIONS,
  NEON_SYMBOLS,
  NEON_ABBREVIATIONS,
  isNeonObject,
  isNeonArray,
  isNeonPrimitive,
} from './types';

// =============================================================================
// Number Compression
// =============================================================================

const NUMBER_SUFFIXES = [
  { threshold: 1_000_000_000_000, suffix: 'T', divisor: 1_000_000_000_000 },
  { threshold: 1_000_000_000, suffix: 'B', divisor: 1_000_000_000 },
  { threshold: 1_000_000, suffix: 'M', divisor: 1_000_000 },
  { threshold: 1_000, suffix: 'K', divisor: 1_000 },
];

function compressNumber(n: number, compress: boolean): string {
  if (!compress || !Number.isFinite(n)) {
    return String(n);
  }

  // Handle negative numbers
  const sign = n < 0 ? '-' : '';
  const absN = Math.abs(n);

  // Very small decimals
  if (absN > 0 && absN < 1) {
    const str = String(n);
    return str.replace(/^(-?)0\./, '$1.');
  }

  // Large numbers with suffix
  for (const { threshold, suffix, divisor } of NUMBER_SUFFIXES) {
    if (absN >= threshold) {
      const divided = absN / divisor;
      // Use up to 2 decimal places, remove trailing zeros
      const formatted = divided.toFixed(2).replace(/\.?0+$/, '');
      return sign + formatted + suffix;
    }
  }

  // Integer check - remove unnecessary decimal
  if (Number.isInteger(n)) {
    return String(n);
  }

  // Regular decimal - limit precision to avoid floating point issues
  const str = n.toPrecision(15).replace(/\.?0+$/, '');
  return str;
}

function expandNumber(s: string): number {
  const lastChar = s.slice(-1).toUpperCase();
  const multipliers: Record<string, number> = {
    'K': 1_000,
    'M': 1_000_000,
    'B': 1_000_000_000,
    'T': 1_000_000_000_000,
  };

  if (multipliers[lastChar]) {
    return parseFloat(s.slice(0, -1)) * multipliers[lastChar];
  }

  if (s.startsWith('.') || s.startsWith('-.')) {
    return parseFloat(s.replace(/^(-?)\./, '$10.'));
  }

  return parseFloat(s);
}

// =============================================================================
// String Compression
// =============================================================================

const COMMON_DOMAINS = [
  { full: '@gmail.com', short: '@gm' },
  { full: '@yahoo.com', short: '@yh' },
  { full: '@hotmail.com', short: '@hm' },
  { full: '@outlook.com', short: '@ol' },
  { full: '@company.com', short: '@co' },
  { full: '@example.com', short: '@ex' },
];

function compressString(s: string, compress: boolean, abbreviations: Map<string, string>): string {
  if (!compress || !s) return s;

  // Check abbreviation map first
  if (abbreviations.has(s)) {
    return abbreviations.get(s)!;
  }

  // Check built-in abbreviations
  if (NEON_ABBREVIATIONS[s]) {
    return NEON_ABBREVIATIONS[s];
  }

  // Compress email domains
  for (const { full, short } of COMMON_DOMAINS) {
    if (s.endsWith(full)) {
      return s.slice(0, -full.length) + short;
    }
  }

  return s;
}

// =============================================================================
// Schema Detection & Generation
// =============================================================================

function detectSchema(arr: NeonObject[]): NeonSchema | null {
  if (arr.length === 0) return null;

  const first = arr[0];
  if (!isNeonObject(first)) return null;

  const keys = Object.keys(first);
  if (keys.length === 0) return null;

  // Check if all objects have the same keys
  const isUniform = arr.every(item => {
    if (!isNeonObject(item)) return false;
    const itemKeys = Object.keys(item);
    return itemKeys.length === keys.length &&
           keys.every(k => k in item);
  });

  if (!isUniform) return null;

  // Generate schema fields with type inference
  const fields: NeonSchemaField[] = keys.map(key => {
    const values = arr.map(item => (item as NeonObject)[key]);
    const type = inferFieldType(values);
    return { name: key, type };
  });

  return { fields };
}

function inferFieldType(values: NeonValue[]): NeonFieldType {
  const types = new Set<NeonFieldType>();

  for (const v of values) {
    if (v === null) continue;
    if (typeof v === 'boolean') types.add('boolean');
    else if (typeof v === 'number') {
      types.add(Number.isInteger(v) ? 'int' : 'float');
    }
    else if (typeof v === 'string') {
      // Try to detect specific string types
      if (/^\d{4}-\d{2}-\d{2}$/.test(v)) types.add('date');
      else if (/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/.test(v)) types.add('datetime');
      else if (/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(v)) types.add('email');
      else if (/^https?:\/\//.test(v)) types.add('url');
      else if (/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(v)) types.add('uuid');
      else types.add('string');
    }
    else if (Array.isArray(v)) types.add('array');
    else if (typeof v === 'object') types.add('object');
  }

  if (types.size === 0) return 'null';
  if (types.size === 1) {
    const firstType = types.values().next().value;
    return firstType ?? 'any';
  }
  if (types.has('int') && types.has('float')) return 'number';
  return 'any';
}

// =============================================================================
// Main Encoder Class
// =============================================================================

export class NeonEncoder {
  private options: Required<Omit<NeonEncodeOptions, 'abbreviationMap' | 'schemaRegistry' | 'flushCallback'>>;
  private abbreviations: Map<string, string>;
  private referenceMap: Map<NeonValue, string>;
  private referenceCounter: number;
  private stats: Partial<NeonStats>;

  constructor(options: NeonEncodeOptions = {}) {
    this.options = { ...DEFAULT_ENCODE_OPTIONS, ...options };
    this.abbreviations = options.abbreviationMap || new Map();
    this.referenceMap = new Map();
    this.referenceCounter = 0;
    this.stats = {};
  }

  /**
   * Encode any JavaScript value to NEON format
   */
  encode(value: NeonValue): string {
    const startTime = performance.now();

    // Reset state
    this.referenceMap = new Map();
    this.referenceCounter = 0;

    // First pass: detect references if enabled
    if (this.options.enableReferences) {
      this.detectReferences(value);
    }

    // Second pass: encode
    const result = this.encodeValue(value, 0, '');

    const endTime = performance.now();

    // Calculate stats
    const jsonStr = JSON.stringify(value);
    this.stats = {
      inputSize: jsonStr.length,
      outputSize: result.length,
      compressionRatio: result.length / jsonStr.length,
      savingsPercent: Math.round((1 - result.length / jsonStr.length) * 100),
      inputTokens: Math.ceil(jsonStr.length / 4),
      outputTokens: Math.ceil(result.length / 4),
      encodeTimeMs: endTime - startTime,
    };

    return result;
  }

  /**
   * Get encoding statistics from last encode operation
   */
  getStats(): Partial<NeonStats> {
    return { ...this.stats };
  }

  /**
   * Detect values that appear multiple times for reference optimization
   */
  private detectReferences(value: NeonValue, seen = new Map<any, number>()): void {
    if (value === null || typeof value !== 'object') return;

    const existing = seen.get(value);
    if (existing !== undefined) {
      seen.set(value, existing + 1);
      return;
    }

    seen.set(value, 1);

    if (Array.isArray(value)) {
      for (const item of value) {
        this.detectReferences(item, seen);
      }
    } else {
      for (const v of Object.values(value)) {
        this.detectReferences(v, seen);
      }
    }

    // Create references for frequently used objects
    for (const [obj, count] of seen) {
      if (count >= this.options.referenceThreshold && typeof obj === 'object') {
        if (!this.referenceMap.has(obj)) {
          this.referenceMap.set(obj, String(++this.referenceCounter));
        }
      }
    }
  }

  /**
   * Main encoding dispatcher
   */
  private encodeValue(value: NeonValue, depth: number, _context: string): string {
    // Null
    if (value === null || value === undefined) {
      return this.options.compressNulls ? NEON_SYMBOLS.NULL : 'null';
    }

    // Boolean
    if (typeof value === 'boolean') {
      if (this.options.compressBooleans) {
        return value ? NEON_SYMBOLS.TRUE : NEON_SYMBOLS.FALSE;
      }
      return String(value);
    }

    // Number
    if (typeof value === 'number') {
      return compressNumber(value, this.options.compressNumbers);
    }

    // String
    if (typeof value === 'string') {
      return this.encodeString(value);
    }

    // Array
    if (Array.isArray(value)) {
      return this.encodeArray(value, depth);
    }

    // Object
    if (typeof value === 'object') {
      return this.encodeObject(value as NeonObject, depth);
    }

    // Fallback
    return String(value);
  }

  /**
   * Encode string with proper escaping and compression
   */
  private encodeString(s: string): string {
    if (s === '') return '""';

    // Apply compression
    const compressed = compressString(
      s,
      this.options.compressStrings,
      this.abbreviations
    );

    // Check if quoting is needed
    const needsQuotes = this.needsQuotes(compressed);

    if (!needsQuotes) {
      // Replace spaces with underscores for readability
      return compressed.replace(/ /g, '_');
    }

    // Escape and quote
    const escaped = compressed
      .replace(/\\/g, '\\\\')
      .replace(/"/g, '\\"')
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/\t/g, '\\t');

    return `"${escaped}"`;
  }

  /**
   * Encode an object key name (handles spaces and special characters)
   */
  private encodeKeyName(key: string): string {
    if (!key) return '""';

    // Check if key needs quoting
    const needsQuotes =
      key.includes(':') ||
      key.includes('"') ||
      key.includes('\\') ||
      key.includes('\n') ||
      key.startsWith(' ') ||
      key.endsWith(' ') ||
      key === NEON_SYMBOLS.TRUE ||
      key === NEON_SYMBOLS.FALSE ||
      key === NEON_SYMBOLS.NULL ||
      /^-?\d+(\.\d+)?[KMBT]?$/.test(key) ||
      /^[#@$~^>-]/.test(key);

    if (!needsQuotes) {
      // Replace spaces with underscores
      return key.replace(/ /g, '_');
    }

    // Escape and quote
    const escaped = key
      .replace(/\\/g, '\\\\')
      .replace(/"/g, '\\"')
      .replace(/\n/g, '\\n');

    return `"${escaped}"`;
  }

  /**
   * Check if a string needs to be quoted
   * Note: Spaces in the middle of strings are converted to underscores,
   * so they don't require quoting. Only leading/trailing spaces require quotes.
   */
  private needsQuotes(s: string): boolean {
    const del = this.options.delimiter;

    // Check for delimiter if it's not a space (spaces are handled via underscore replacement)
    if (del !== ' ' && s.includes(del)) {
      return true;
    }

    return (
      s.includes(':') ||
      s.includes('"') ||
      s.includes('\\') ||
      s.includes('\n') ||
      s.includes('\r') ||
      s.includes('\t') ||
      s.startsWith(' ') ||
      s.endsWith(' ') ||
      s === NEON_SYMBOLS.TRUE ||
      s === NEON_SYMBOLS.FALSE ||
      s === NEON_SYMBOLS.NULL ||
      /^-?\d+(\.\d+)?[KMBT]?$/.test(s) ||
      /^[#@$~^>-]/.test(s)
    );
  }

  /**
   * Encode array with automatic tabular detection
   */
  private encodeArray(arr: NeonArray, depth: number): string {
    if (arr.length === 0) {
      return `${NEON_SYMBOLS.ARRAY}0`;
    }

    // Try tabular encoding for arrays of uniform objects
    const schema = detectSchema(arr as NeonObject[]);
    if (schema) {
      return this.encodeTabularArray(arr as NeonObject[], schema, depth);
    }

    // Check for primitive array
    const isPrimitive = arr.every(isNeonPrimitive);
    if (isPrimitive) {
      return this.encodePrimitiveArray(arr, depth);
    }

    // Mixed array - use list format
    return this.encodeListArray(arr, depth);
  }

  /**
   * Encode tabular array (uniform objects) - THE SWEET SPOT
   */
  private encodeTabularArray(arr: NeonObject[], schema: NeonSchema, depth: number): string {
    const fields = schema.fields.map(f => f.name);
    const del = this.options.delimiter;
    const nl = this.options.lineEnding;
    const indent = this.getIndent(depth + 1);

    // Abbreviate field names if enabled
    const schemaFields = this.options.abbreviateFields
      ? fields.map(f => NEON_ABBREVIATIONS[f] || f)
      : fields;

    // Header line
    let result = `${NEON_SYMBOLS.ARRAY}${arr.length}${NEON_SYMBOLS.SCHEMA}${schemaFields.join(',')}`;

    // Data rows
    for (const item of arr) {
      result += nl + indent;
      const values = fields.map(f => this.encodeValue(item[f], depth + 1, f));
      result += values.join(del);
    }

    return result;
  }

  /**
   * Encode primitive array on single line
   */
  private encodePrimitiveArray(arr: NeonArray, depth: number): string {
    const del = this.options.delimiter;
    const values = arr.map(v => this.encodeValue(v, depth, ''));
    return `${NEON_SYMBOLS.ARRAY}${arr.length}${del}${values.join(del)}`;
  }

  /**
   * Encode mixed array with list items
   */
  private encodeListArray(arr: NeonArray, depth: number): string {
    const nl = this.options.lineEnding;
    const indent = this.getIndent(depth + 1);

    let result = `${NEON_SYMBOLS.ARRAY}${arr.length}`;

    for (const item of arr) {
      result += nl + indent + '- ';
      const encoded = this.encodeValue(item, depth + 1, '');
      // Handle multi-line values
      if (encoded.includes('\n')) {
        result += encoded.split('\n').join('\n' + indent + '  ');
      } else {
        result += encoded;
      }
    }

    return result;
  }

  /**
   * Encode object
   */
  private encodeObject(obj: NeonObject, depth: number): string {
    const entries = Object.entries(obj);

    if (entries.length === 0) {
      return NEON_SYMBOLS.OBJECT;
    }

    // Check if this is a root object with a single array property (common case)
    if (depth === 0 && entries.length === 1) {
      const [key, value] = entries[0];
      if (Array.isArray(value)) {
        const schema = detectSchema(value as NeonObject[]);
        if (schema) {
          // Encode as named tabular array
          const fields = schema.fields.map(f => f.name);
          const schemaFields = this.options.abbreviateFields
            ? fields.map(f => NEON_ABBREVIATIONS[f] || f)
            : fields;

          const del = this.options.delimiter;
          const nl = this.options.lineEnding;
          const indent = this.getIndent(1);

          let result = `${key}${NEON_SYMBOLS.ARRAY}${value.length}${NEON_SYMBOLS.SCHEMA}${schemaFields.join(',')}`;

          for (const item of value as NeonObject[]) {
            result += nl + indent;
            const values = fields.map(f => this.encodeValue(item[f], 1, f));
            result += values.join(del);
          }

          return result;
        }
      }
    }

    // Standard object encoding
    const parts: string[] = [];
    const nl = this.options.lineEnding;
    const childIndent = this.getIndent(depth + 1);

    for (const [key, value] of entries) {
      const abbreviatedKey = this.options.abbreviateFields
        ? (NEON_ABBREVIATIONS[key] || key)
        : key;
      const encodedKey = this.encodeKeyName(abbreviatedKey);

      if (isNeonObject(value) && Object.keys(value).length > 0) {
        // Nested object
        const nested = this.encodeObject(value, depth + 1);
        if (this.options.mode === 'compact' || this.options.mode === 'ultra-compact') {
          parts.push(`${encodedKey}:{${nested.slice(1)}}`);
        } else {
          parts.push(`${encodedKey}:${nl}${childIndent}${nested}`);
        }
      } else if (isNeonArray(value)) {
        // Array property
        const encoded = this.encodeArray(value, depth + 1);
        if (encoded.includes('\n')) {
          parts.push(`${encodedKey}${encoded}`);
        } else {
          parts.push(`${encodedKey}${encoded}`);
        }
      } else {
        // Primitive
        const encoded = this.encodeValue(value, depth, key);
        parts.push(`${encodedKey}:${encoded}`);
      }
    }

    if (this.options.mode === 'compact' || this.options.mode === 'ultra-compact') {
      return NEON_SYMBOLS.OBJECT + parts.join(' ');
    }

    return NEON_SYMBOLS.OBJECT + parts.join(' ');
  }

  /**
   * Get indentation string for given depth
   */
  private getIndent(depth: number): string {
    if (this.options.mode === 'ultra-compact') return '';
    if (this.options.mode === 'compact') return '';
    return ' '.repeat(this.options.indent * depth);
  }
}

// =============================================================================
// Convenience Functions
// =============================================================================

/**
 * Encode value to NEON format with default options
 */
export function encode(value: NeonValue, options?: NeonEncodeOptions): string {
  const encoder = new NeonEncoder(options);
  return encoder.encode(value);
}

/**
 * Encode with compact mode
 */
export function encodeCompact(value: NeonValue): string {
  return encode(value, { mode: 'compact' });
}

/**
 * Encode with ultra-compact mode for maximum compression
 */
export function encodeUltraCompact(value: NeonValue): string {
  return encode(value, { mode: 'ultra-compact' });
}

/**
 * Encode with readable mode for human inspection
 */
export function encodeReadable(value: NeonValue): string {
  return encode(value, { mode: 'readable', indent: 2 });
}

/**
 * Get encoding statistics without returning the encoded string
 */
export function getEncodingStats(value: NeonValue, options?: NeonEncodeOptions): NeonStats {
  const encoder = new NeonEncoder(options);
  const encoded = encoder.encode(value);
  const stats = encoder.getStats();

  const json = JSON.stringify(value);

  return {
    inputSize: json.length,
    outputSize: encoded.length,
    compressionRatio: encoded.length / json.length,
    savingsPercent: Math.round((1 - encoded.length / json.length) * 100),
    inputTokens: Math.ceil(json.length / 4),
    outputTokens: Math.ceil(encoded.length / 4),
    tokenSavings: Math.ceil(json.length / 4) - Math.ceil(encoded.length / 4),
    objectCount: 0, // Would need traversal to calculate
    arrayCount: 0,
    primitiveCount: 0,
    maxDepth: 0,
    encodeTimeMs: stats.encodeTimeMs,
    estimatedCostJson: (Math.ceil(json.length / 4) / 1000) * 0.01,
    estimatedCostNeon: (Math.ceil(encoded.length / 4) / 1000) * 0.01,
    costSavings: ((Math.ceil(json.length / 4) - Math.ceil(encoded.length / 4)) / 1000) * 0.01,
  };
}

// Export number utilities for decoder
export { expandNumber, compressNumber };

File: rust/src/encoder.rs
//! NEON Encoder Implementation

use crate::error::Result;
use crate::types::{symbols, get_abbreviations, NeonEncodeOptions, NeonStats};
use serde_json::Value;
use std::time::Instant;

/// Compress a number using K/M/B/T suffixes
fn compress_number(n: f64, enabled: bool) -> String {
    if !enabled {
        if n.fract() == 0.0 {
            return format!("{}", n as i64);
        }
        return format!("{}", n);
    }

    let abs_n = n.abs();
    let sign = if n < 0.0 { "-" } else { "" };

    if abs_n >= 1_000_000_000_000.0 {
        let val = abs_n / 1_000_000_000_000.0;
        if val.fract() == 0.0 {
            return format!("{}{}T", sign, val as i64);
        }
        return format!("{}{:.1}T", sign, val);
    }
    if abs_n >= 1_000_000_000.0 {
        let val = abs_n / 1_000_000_000.0;
        if val.fract() == 0.0 {
            return format!("{}{}B", sign, val as i64);
        }
        return format!("{}{:.1}B", sign, val);
    }
    if abs_n >= 1_000_000.0 {
        let val = abs_n / 1_000_000.0;
        if val.fract() == 0.0 {
            return format!("{}{}M", sign, val as i64);
        }
        return format!("{}{:.1}M", sign, val);
    }
    if abs_n >= 1_000.0 {
        let val = abs_n / 1_000.0;
        if val.fract() == 0.0 {
            return format!("{}{}K", sign, val as i64);
        }
        return format!("{}{:.1}K", sign, val);
    }

    // Leading dot for small decimals
    if abs_n > 0.0 && abs_n < 1.0 {
        let s = format!("{}", n);
        return s.replace("0.", ".");
    }

    // Integer check
    if n.fract() == 0.0 {
        return format!("{}", n as i64);
    }

    format!("{}", n)
}

/// Check if a string needs to be quoted
fn needs_quotes(s: &str, delimiter: char) -> bool {
    if s.is_empty() {
        return true;
    }

    let special_chars: Vec<char> = vec![':', '"', '\\', '\n', '\r', '\t'];

    if special_chars.iter().any(|c| s.contains(*c)) {
        return true;
    }
    if delimiter != ' ' && s.contains(delimiter) {
        return true;
    }
    if s.starts_with(' ') || s.ends_with(' ') {
        return true;
    }
    if s == "T" || s == "F" || s == "N" {
        return true;
    }

    // Check if looks like a number
    let num_pattern = regex_lite::Regex::new(r"^-?\d+(\.\d+)?[KMBT]?$").unwrap();
    if num_pattern.is_match(s) {
        return true;
    }

    // Check if starts with special character
    if let Some(first) = s.chars().next() {
        if "#@$~^>-".contains(first) {
            return true;
        }
    }

    false
}

/// Escape special characters in a string
fn escape_string(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
        .replace('\r', "\\r")
        .replace('\t', "\\t")
}

/// NEON Encoder
pub struct NeonEncoder {
    options: NeonEncodeOptions,
    stats: NeonStats,
}

impl NeonEncoder {
    pub fn new(options: Option<NeonEncodeOptions>) -> Self {
        Self {
            options: options.unwrap_or_default(),
            stats: NeonStats::default(),
        }
    }

    pub fn encode(&mut self, value: &Value) -> Result<String> {
        let start = Instant::now();

        let result = self.encode_value(value, 0)?;

        let elapsed = start.elapsed();
        let json_str = serde_json::to_string(value)?;

        self.stats = NeonStats {
            input_size: json_str.len(),
            output_size: result.len(),
            compression_ratio: result.len() as f64 / json_str.len() as f64,
            savings_percent: ((1.0 - result.len() as f64 / json_str.len() as f64) * 100.0) as i32,
            input_tokens: json_str.len() / 4,
            output_tokens: result.len() / 4,
            encode_time_ms: elapsed.as_secs_f64() * 1000.0,
            ..Default::default()
        };

        Ok(result)
    }

    pub fn get_stats(&self) -> &NeonStats {
        &self.stats
    }

    fn encode_value(&self, value: &Value, depth: usize) -> Result<String> {
        match value {
            Value::Null => {
                if self.options.compress_nulls {
                    Ok(symbols::NULL.to_string())
                } else {
                    Ok("null".to_string())
                }
            }
            Value::Bool(b) => {
                if self.options.compress_booleans {
                    Ok(if *b {
                        symbols::TRUE.to_string()
                    } else {
                        symbols::FALSE.to_string()
                    })
                } else {
                    Ok(b.to_string())
                }
            }
            Value::Number(n) => {
                let num = n.as_f64().unwrap_or(0.0);
                Ok(compress_number(num, self.options.compress_numbers))
            }
            Value::String(s) => Ok(self.encode_string(s)),
            Value::Array(arr) => self.encode_array(arr, depth),
            Value::Object(obj) => self.encode_object(obj, depth),
        }
    }

    fn encode_string(&self, s: &str) -> String {
        if s.is_empty() {
            return "\"\"".to_string();
        }

        if needs_quotes(s, self.options.delimiter) {
            return format!("\"{}\"", escape_string(s));
        }

        s.replace(' ', "_")
    }

    fn encode_key(&self, key: &str) -> String {
        if key.is_empty() {
            return "\"\"".to_string();
        }

        let key = if self.options.abbreviate_fields {
            get_abbreviations()
                .get(key)
                .map(|s| *s)
                .unwrap_or(key)
        } else {
            key
        };

        if key.contains(':') || key.contains('"') || key.contains('\\') || key.contains('\n') {
            return format!("\"{}\"", escape_string(key));
        }

        key.replace(' ', "_")
    }

    fn encode_array(&self, arr: &[Value], depth: usize) -> Result<String> {
        if arr.is_empty() {
            return Ok(format!("{}0", symbols::ARRAY));
        }

        // Check if tabular
        if self.is_tabular(arr) {
            return self.encode_tabular_array(arr, depth);
        }

        // Check if primitive
        let is_primitive = arr.iter().all(|v| !v.is_array() && !v.is_object());
        if is_primitive {
            return self.encode_primitive_array(arr);
        }

        // Mixed array
        self.encode_list_array(arr, depth)
    }

    fn is_tabular(&self, arr: &[Value]) -> bool {
        if arr.is_empty() {
            return false;
        }

        let first = match &arr[0] {
            Value::Object(obj) => obj,
            _ => return false,
        };

        let first_keys: std::collections::HashSet<_> = first.keys().collect();

        arr.iter().all(|v| {
            if let Value::Object(obj) = v {
                let keys: std::collections::HashSet<_> = obj.keys().collect();
                keys == first_keys
            } else {
                false
            }
        })
    }

    fn encode_tabular_array(&self, arr: &[Value], depth: usize) -> Result<String> {
        let first = arr[0].as_object().unwrap();
        let fields: Vec<&String> = first.keys().collect();

        let schema_fields: Vec<String> = fields
            .iter()
            .map(|f| {
                if self.options.abbreviate_fields {
                    get_abbreviations()
                        .get(f.as_str())
                        .map(|s| s.to_string())
                        .unwrap_or_else(|| f.to_string())
                } else {
                    f.to_string()
                }
            })
            .collect();

        let mut result = format!(
            "{}{}{}{}",
            symbols::ARRAY,
            arr.len(),
            symbols::SCHEMA,
            schema_fields.join(",")
        );

        let indent = " ".repeat(self.options.indent * (depth + 1));

        for item in arr {
            if let Value::Object(obj) = item {
                let values: Vec<String> = fields
                    .iter()
                    .map(|f| {
                        obj.get(*f)
                            .map(|v| self.encode_value(v, depth + 1).unwrap_or_default())
                            .unwrap_or_default()
                    })
                    .collect();

                result.push_str(&self.options.line_ending);
                result.push_str(&indent);
                result.push_str(&values.join(&self.options.delimiter.to_string()));
            }
        }

        Ok(result)
    }

    fn encode_primitive_array(&self, arr: &[Value]) -> Result<String> {
        let values: Vec<String> = arr
            .iter()
            .map(|v| self.encode_value(v, 0).unwrap_or_default())
            .collect();

        Ok(format!(
            "{}{} {}",
            symbols::ARRAY,
            arr.len(),
            values.join(&self.options.delimiter.to_string())
        ))
    }

    fn encode_list_array(&self, arr: &[Value], depth: usize) -> Result<String> {
        let indent = " ".repeat(self.options.indent * (depth + 1));
        let mut result = format!("{}{}", symbols::ARRAY, arr.len());

        for item in arr {
            let encoded = self.encode_value(item, depth + 1)?;
            result.push_str(&self.options.line_ending);
            result.push_str(&indent);
            result.push(symbols::LIST_ITEM);
            result.push(' ');
            result.push_str(&encoded);
        }

        Ok(result)
    }

    fn encode_object(
        &self,
        obj: &serde_json::Map<String, Value>,
        depth: usize,
    ) -> Result<String> {
        if obj.is_empty() {
            return Ok(symbols::OBJECT.to_string());
        }

        let entries: Vec<(&String, &Value)> = obj.iter().collect();

        // Check for root object with single array property
        if depth == 0 && entries.len() == 1 {
            let (key, value) = entries[0];
            if let Value::Array(arr) = value {
                if self.is_tabular(arr) {
                    let first = arr[0].as_object().unwrap();
                    let fields: Vec<&String> = first.keys().collect();

                    let schema_fields: Vec<String> = fields
                        .iter()
                        .map(|f| {
                            if self.options.abbreviate_fields {
                                get_abbreviations()
                                    .get(f.as_str())
                                    .map(|s| s.to_string())
                                    .unwrap_or_else(|| f.to_string())
                            } else {
                                f.to_string()
                            }
                        })
                        .collect();

                    let mut result = format!(
                        "{}{}{}{}{}",
                        key,
                        symbols::ARRAY,
                        arr.len(),
                        symbols::SCHEMA,
                        schema_fields.join(",")
                    );

                    let indent = " ".repeat(self.options.indent);

                    for item in arr {
                        if let Value::Object(item_obj) = item {
                            let values: Vec<String> = fields
                                .iter()
                                .map(|f| {
                                    item_obj
                                        .get(*f)
                                        .map(|v| self.encode_value(v, 1).unwrap_or_default())
                                        .unwrap_or_default()
                                })
                                .collect();

                            result.push_str(&self.options.line_ending);
                            result.push_str(&indent);
                            result.push_str(&values.join(&self.options.delimiter.to_string()));
                        }
                    }

                    return Ok(result);
                }
            }
        }

        // Standard object encoding
        let mut parts: Vec<String> = Vec::new();

        for (key, value) in entries {
            let encoded_key = self.encode_key(key);

            match value {
                Value::Object(nested) if !nested.is_empty() => {
                    let nested_encoded = self.encode_object(nested, depth + 1)?;
                    parts.push(format!("{}:{{{}}}", encoded_key, &nested_encoded[1..]));
                }
                Value::Array(arr) => {
                    let encoded = self.encode_array(arr, depth + 1)?;
                    parts.push(format!("{}{}", encoded_key, encoded));
                }
                _ => {
                    let encoded = self.encode_value(value, depth)?;
                    parts.push(format!("{}:{}", encoded_key, encoded));
                }
            }
        }

        Ok(format!("{}{}", symbols::OBJECT, parts.join(" ")))
    }
}

/// Encode a JSON value to NEON format
pub fn encode(value: &Value, options: Option<NeonEncodeOptions>) -> Result<String> {
    let mut encoder = NeonEncoder::new(options);
    encoder.encode(value)
}

/// Encode with maximum compression
pub fn encode_compact(value: &Value) -> Result<String> {
    let options = NeonEncodeOptions {
        compress_numbers: true,
        compress_booleans: true,
        compress_nulls: true,
        abbreviate_fields: true,
        ..Default::default()
    };
    encode(value, Some(options))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_encode_null() {
        let result = encode(&json!(null), None).unwrap();
        assert_eq!(result, "N");
    }

    #[test]
    fn test_encode_boolean() {
        assert_eq!(encode(&json!(true), None).unwrap(), "T");
        assert_eq!(encode(&json!(false), None).unwrap(), "F");
    }

    #[test]
    fn test_encode_number() {
        assert_eq!(encode(&json!(42), None).unwrap(), "42");
        assert_eq!(encode(&json!(95000), None).unwrap(), "95K");
        assert_eq!(encode(&json!(2500000), None).unwrap(), "2.5M");
    }

    #[test]
    fn test_encode_string() {
        assert_eq!(encode(&json!("hello"), None).unwrap(), "hello");
        assert_eq!(encode(&json!("hello world"), None).unwrap(), "hello_world");
    }

    #[test]
    fn test_encode_array() {
        let result = encode(&json!([1, 2, 3]), None).unwrap();
        assert_eq!(result, "#3 1 2 3");
    }

    #[test]
    fn test_encode_tabular() {
        let data = json!({
            "users": [
                {"id": 1, "name": "Alice"},
                {"id": 2, "name": "Bob"}
            ]
        });
        let result = encode(&data, None).unwrap();
        assert!(result.contains("users#2^"));
        assert!(result.contains("Alice"));
        assert!(result.contains("Bob"));
    }
}
